pgrx/src/list.rs

```diff
@@ -12,257 +12,617 @@
+        /// From a non-nullable pointer that points to a valid List, produce a ListHead of the correct type
+        ///
+        /// # Safety
+        /// This assumes the NodeTag is valid to read, so it is not okay to call this on
+        /// pointers to deallocated or uninit data.
+        ///
+        /// If it returns as `Some`, it also asserts the entire List is, across its length,
+        /// validly initialized as `T` in each ListCell.
+        pub unsafe fn downcast_from_ptr(list: NonNull<pg_sys::List>) -> Option<ListHead<T>> {
+            (T::LIST_TAG == (*list.as_ptr()).type_).then_some(ListHead { list, _type: PhantomData })
+        }
+
+        pub fn push(&mut self, value: T) -> &mut Self {
+            let list = unsafe { self.list.as_mut() };
+            let pg_sys::List { length, max_length, elements, .. } = list;
+            if *max_length - *length > 0 {
+                // SAFETY: Our list must have been constructed following the list invariants
+                // in order to actually get here, and we have confirmed as in-range of the buffer.
+                let cell = unsafe { &mut *elements.add(*length as _) };
+                T::endocytosis(cell, value);
+                *length += 1;
+            } else {
+                // Reserve in this branch.
+                let new_cap = max_length.saturating_mul(2);
+                self.reserve(new_cap as _);
+            }
+
+            // Return `self` for convenience of `List::try_push`
+            self
+        }
+
+        pub fn reserve(&mut self, size: usize) -> &mut Self {
+            let list = unsafe { self.list.as_mut() };
+            if ((list.max_length - list.length) as usize) < size {
+                unsafe { grow_list(list, size + list.length as usize) };
+            };
+            self
         }
     }
 
-    #[inline]
-    pub fn iter_ptr(&self) -> impl Iterator<Item = *mut T> + '_ {
-        PgListIteratorPtr { list: &self, pos: 0 }
+    unsafe fn grow_list(list: &mut pg_sys::List, target: usize) {
+        let alloc_size = target * mem::size_of::<pg_sys::ListCell>();
+        if list.elements == ptr::addr_of_mut!(list.initial_elements).cast() {
+            // first realloc, we can't dealloc the elements ptr, as it isn't its own alloc
+            let context = pg_sys::GetMemoryChunkContext(list as *mut _ as *mut _);
+            if context == ptr::null_mut() {
+                panic!("Context free list?");
+            }
+            let buf = pg_sys::MemoryContextAlloc(context, alloc_size);
+            if buf == ptr::null_mut() {
+                panic!("List allocation failure");
+            }
+            ptr::copy_nonoverlapping(list.elements, buf.cast(), list.length as _);
+            // If the old buffer is pointers, we would like everyone dereferencing them to segfault,
+            // if OIDs, Postgres will surface errors quickly on InvalidOid, etc.
+            ptr::write_bytes(list.elements, 0, list.length as _);
```

Comment:
However, Postgres only does this when `CLOBBER_FREED_MEMORY` is defined.  How faithful to their version do we want to be?  I suspect they don't want it in --release builds and maybe we don't either?  

For reference: https://github.com/postgres/postgres/blob/a899d07b24a6e519790041eeea5d96cc534f384e/src/backend/nodes/list.c#L196

Reply:
I would prefer to have one consistent approach throughout, basically. I think a short memset is unlikely to be a performance concern.

Reply:
basically the reason to always do it is so we don't have to bind the valgrind headers, and we may still surface some bugs this way, because they use VALGRIND_MAKE_MEM_NOACCESS.

Reply:
Don't clobber memory on release builds, probably.

Reply:
I corrected the clobber pattern and made it a `#[cfg(debug_assertions)]` thing!